packages/lesswrong/server/resolvers/bookmarksFeedResolver.ts

```diff
@@ -0,0 +1,165 @@
+import gql from 'graphql-tag';
+import { Bookmarks } from '../collections/bookmarks/collection';
+import { loadByIds } from '../../lib/loaders';
+import { filterNonnull } from '../../lib/utils/typeGuardUtils';
+import { FeedPostMetaInfo, FeedCommentMetaInfo, FeedPostResolverType, FeedCommentsThreadResolverType as UltraFeedCommentsThreadResolverType } from '../../components/ultraFeed/ultraFeedTypes';
+import uniq from 'lodash/uniq';
+import orderBy from 'lodash/orderBy';
+
+export const bookmarksFeedGraphQLTypeDefs = gql`
+  type BookmarksFeedQueryResults {
+    cutoff: Date
+    endOffset: Int!
+    results: [BookmarksFeedEntryType!]
+  }
+  
+  type BookmarksFeedEntryType {
+    type: String!          # 'feedPost' or 'feedCommentThread'
+    feedPost: FeedPost     
+    feedCommentThread: FeedCommentThread 
+    sortKey: Date!         
+  }
+  
+  extend type Query {
+    BookmarksFeed(
+      limit: Int,
+      cutoff: Date,
+      offset: Int,
+    ): BookmarksFeedQueryResults!
+  }
+`
+
+interface FeedCommentsThreadResolverType extends UltraFeedCommentsThreadResolverType {
+  post: Partial<DbPost> | null;
+}
+
+type InternalFeedItem = {
+  type: 'feedPost' | 'feedCommentThread';
+  feedPost: FeedPostResolverType | null;
+  feedCommentThread: FeedCommentsThreadResolverType | null;
+  sortKey: Date;
+};
+
+interface BookmarksFeedQueryResult {
+  cutoff: Date | null;
+  endOffset: number;
+  results: BookmarksFeedEntryType[];
+}
+
+export const bookmarksFeedGraphQLQueries = {
+  BookmarksFeed: async (_root: void, args: any, context: ResolverContext): Promise<BookmarksFeedQueryResult> => {
+    const {limit = 20, cutoff, offset = 0} = args;
+    const {currentUser} = context;
+    
+    if (!currentUser) {
+      throw new Error("Must be logged in to view bookmarks feed");
+    }
+
+    const selector: MongoSelector<DbBookmark> = {
+      userId: currentUser._id,
+      cancelled: false,
+    };
+    if (cutoff) {
+      selector.lastUpdated = {$lt: cutoff}; 
+    }
+
+    const bookmarks = await Bookmarks.find(selector, {
+      sort: {lastUpdated: -1},
```

Comment:
The selector and sort for this query imply an index on `userId, active [formerly cancelled], lastUpdated`, since you aren't sorting on `createdAt`.